### PR TITLE
Refactoring into Visitor pattern

### PR DESCRIPTION
**Pull request by HosseinYousefi**
This refactor will make the code for Dart binding generators more uniform.

Introducing a top-level `Classes` class which includes all the classes we're trying to generate as yet another element, makes each step of the code generation pipeline, feel the same. For example exclusion of classes and methods are now both done via actually removing the objects instead of setting an `isIncluded` flag. 

This `Classes` object will go through a pipeline of:

1. `Excluder` – excludes the unwanted classes, the private ones and the ones specified by the user.
2. `Linker` – links up the objects together to create a graph.
3. `Renamer` – renames the classes, methods, fields, ... not to clash with other reserved names.
4. `DartGenerator` – generates and writes to files.

If the output mode is set to C-based, then a C binding generator will also be run. I haven't yet refactored the C binding generation code to also use visitor pattern. This can happen in the future. However it's not a top priority as the code only affects the C generation and we don't have multiple kinds of C generation code.

This is the first step towards making the code even more pipeline like, so we can test each step of the way better. There are some changes that could be done to the APISummarizer too. However it would take more time and make the review of this PR even more difficult to do so I decided against it for the time being.

### TODO:

- [x] Change the relevant architecture & design section in wiki.

### CODE REVIEW

File: jnigen/lib/src/bindings/dart_generator.dart [hunk 1]
```diff
@@ -0,0 +1,1111 @@
...
+''';
+
+  static const autoGeneratedNotice = '// Autogenerated by jnigen. '
+      'DO NOT EDIT!\n\n';
+  static const defaultImports = 'import "dart:isolate" show ReceivePort;\n'
+      'import "dart:ffi" as ffi;\n'
+      'import "package:jni/internal_helpers_for_jnigen.dart";\n'
+      'import "package:jni/jni.dart" as jni;\n\n';
+  static const defaultLintSuppressions =
+      '// ignore_for_file: annotate_overrides\n'
```
Comment (dcharkes) [inline jnigen/lib/src/bindings/dart_generator.dart:128]: Please use:

```dart
'''
So that we don't have to
write \n's
''';
```

**Review by dcharkes — APPROVED** [review]

File: jnigen/lib/src/bindings/dart_generator.dart [hunk 2]
```diff
@@ -0,0 +1,1111 @@
...
+String _encloseIfNotEmpty(String open, String inside, String close) {
+  if (inside == '') return '';
+  return '$open$inside$close';
+}
+
+String _newLine({int depth = 0}) {
+  return '\n${'  ' * depth}';
+}
+
+// **Naming Convention**
```
Comment (dcharkes) [inline jnigen/lib/src/bindings/dart_generator.dart:66]: ///

File: jnigen/lib/src/bindings/dart_generator.dart [hunk 3]
```diff
@@ -0,0 +1,1111 @@
...
+const _accessors = 'jniAccessors';
+const _lookup = 'jniLookup';
+const _selfPointer = 'reference';
+
+// Docs
+const _deleteInstruction =
+    '  /// The returned object must be deleted after use, '
+    'by calling the `delete` method.';
+
+extension _StringX on String {
```
[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply jnigen/lib/src/bindings/dart_generator.dart:49]: TIL!

File: jnigen/lib/src/bindings/linker.dart [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'visitor.dart';
+import '../config/config.dart';
+import '../elements/elements.dart';
+
+typedef _Resolver = ClassDecl Function(String? binaryName);
+
+/// Adds references from child elements back to their parent elements.
```
[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply jnigen/lib/src/bindings/linker.dart:11]: Not only linker, but renamer and excluder must also be called. `Element`s have `late final` fields. I added comments to the fields specifying exactly at which stage they'll be populated.

File: jnigen/lib/src/bindings/dart_generator.dart (hunk 1 above)
Comment (HosseinYousefi) [reply jnigen/lib/src/bindings/dart_generator.dart:128]: Copied from the old version! Will change!

File: jnigen/lib/src/elements/elements.dart [hunk 5]
```diff
@@ -20,38 +27,32 @@ enum DeclKind {
   enumKind,
 }
 
+class Classes implements Element<Classes> {
```
[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply jnigen/lib/src/elements/elements.dart:30]: This is the representation of the classes the user chose to generate using JNIgen which doesn't have to necessarily be in one module. `Project` felt too generic.

File: jnigen/lib/src/bindings/visitor.dart [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  Iterable<R> accept<R>(TypeVisitor<R> v) {
+    return map((e) => e.accept(v));
+  }
+}
+
+extension MultiTypeUsageVisitor on Iterable<TypeUsage> {
+  Iterable<R> accept<R>(TypeVisitor<R> v) {
+    return map((e) => e.type.accept(v));
+  }
+}
```
[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply jnigen/lib/src/bindings/visitor.dart:40]: Code generation ordering is different case by case. I found it more clear to be specific about which children to visit and in which order.

**Pull request merged**